chia/full_node/full_node.py

```diff
@@ -1111,27 +1112,75 @@ async def sync_from_fork_point(
                 # gymnastics of this range (+1 to make it exclusive, like normal
                 # ranges) and then -1 when forming the request message
                 for start_height in range(fork_point_height, target_peak_sb_height + 1, batch_size):
                     end_height = min(target_peak_sb_height, start_height + batch_size - 1)
                     request = RequestBlocks(uint32(start_height), uint32(end_height), True)
+                    new_peers_with_peak.sort(key=lambda pair: pair[1])
                     fetched = False
-                    for peer in random.sample(new_peers_with_peak, len(new_peers_with_peak)):
+                    for idx, (peer, timestamp) in enumerate(new_peers_with_peak):
                         if peer.closed:
                             continue
+
                         start = time.monotonic()
+                        if start < timestamp:
+                            # rate limit ourselves, since we sent a message to
+                            # this peer too recently
+                            await asyncio.sleep(timestamp - start)
+                            start = time.monotonic()
+
+                        # update the timestamp, now that we're sending a request
+                        if is_localhost(peer.peer_info.host):
+                            # we don't apply rate limits to localhost, and our
+                            # tests depend on it
+                            new_peers_with_peak[idx] = (
+                                new_peers_with_peak[idx][0],
+                                new_peers_with_peak[idx][1] + 0.1,
+                            )
+                        else:
+                            new_peers_with_peak[idx] = (
+                                new_peers_with_peak[idx][0],
+                                new_peers_with_peak[idx][1] + seconds_per_request,
+                            )
                         response = await peer.call_api(FullNodeAPI.request_blocks, request, timeout=30)
                         end = time.monotonic()
-                        if end - start > 5:
-                            self.log.info(f"sync pipeline, peer took {end-start:0.2f} to respond to request_blocks")
                         if response is None:
+                            self.log.info(f"peer timed out after {end-start:.1f} s")
                             await peer.close()
                         elif isinstance(response, RespondBlocks):
+                            if end - start > 5:
+                                self.log.info(f"peer took {end-start:.1f} s to respond to request_blocks")
+                                # this isn't a great peer, reduce its priority
+                                # to prefer any peers that had to wait for it
+                                new_peers_with_peak[idx] = (
+                                    new_peers_with_peak[idx][0],
+                                    end,
```

Comment:
do i understand correctly that we mesure from when the response came and not from when we sent the request ?

Reply:
in the normal case, we just bump the timestamp by 2 seconds (or 0.1 for localhost). So, if we fall behind because one peer is slow, we preserve the space to catch up with more requests. The logic right here is where a peer took more than 5 seconds to respond, we "penalize" it by resetting the timestamp to the current time. i.e. the 5 seconds we just waited, and accrued the opportunity to send requests, we remove from this peer. This will make us prefer other peers over this one for at least 3 requests each (assuming 2 seconds per request).

Since we don't make requests concurrently, but to one peer at a time, this prevents a single slow peer from ruining the sync rate from other peers.